File: app/src/main/java/com/nextcloud/talk/webrtc/MagicWebSocketInstance.java
/*
 * Nextcloud Talk application
 *
 * @author Mario Danic
 * Copyright (C) 2017-2018 Mario Danic <mario@lovelyhq.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.nextcloud.talk.webrtc;

import android.content.Context;
import android.text.TextUtils;
import android.util.Log;

import com.bluelinelabs.logansquare.LoganSquare;
import com.nextcloud.talk.application.NextcloudTalkApplication;
import com.nextcloud.talk.data.user.model.User;
import com.nextcloud.talk.events.NetworkEvent;
import com.nextcloud.talk.events.WebSocketCommunicationEvent;
import com.nextcloud.talk.models.json.participants.Participant;
import com.nextcloud.talk.models.json.signaling.NCSignalingMessage;
import com.nextcloud.talk.models.json.websocket.BaseWebSocketMessage;
import com.nextcloud.talk.models.json.websocket.ByeWebSocketMessage;
import com.nextcloud.talk.models.json.websocket.CallOverallWebSocketMessage;
import com.nextcloud.talk.models.json.websocket.ErrorOverallWebSocketMessage;
import com.nextcloud.talk.models.json.websocket.EventOverallWebSocketMessage;
import com.nextcloud.talk.models.json.websocket.HelloResponseOverallWebSocketMessage;
import com.nextcloud.talk.models.json.websocket.JoinedRoomOverallWebSocketMessage;
import com.nextcloud.talk.signaling.SignalingMessageReceiver;
import com.nextcloud.talk.signaling.SignalingMessageSender;
import com.nextcloud.talk.utils.bundle.BundleKeys;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import autodagger.AutoInjector;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

import static com.nextcloud.talk.models.json.participants.Participant.ActorType.GUESTS;
import static com.nextcloud.talk.models.json.participants.Participant.ActorType.USERS;
import static com.nextcloud.talk.webrtc.Globals.ROOM_TOKEN;
import static com.nextcloud.talk.webrtc.Globals.TARGET_PARTICIPANTS;

@AutoInjector(NextcloudTalkApplication.class)
public class MagicWebSocketInstance extends WebSocketListener {
    private static final String TAG = "MagicWebSocketInstance";


    @Inject
    OkHttpClient okHttpClient;

    @Inject
    EventBus eventBus;

    @Inject
    Context context;

    private User conversationUser;
    private String webSocketTicket;
    private String resumeId;
    private String sessionId;
    private boolean hasMCU;
    private boolean connected;
    private WebSocketConnectionHelper webSocketConnectionHelper;
    private WebSocket internalWebSocket;
    private String connectionUrl;

    private String currentRoomToken;
    private int restartCount = 0;
    private boolean reconnecting = false;

    private HashMap<String, Participant> usersHashMap;

    private List<String> messagesQueue = new ArrayList<>();

    private final ExternalSignalingMessageReceiver signalingMessageReceiver = new ExternalSignalingMessageReceiver();

    private final ExternalSignalingMessageSender signalingMessageSender = new ExternalSignalingMessageSender();

    MagicWebSocketInstance(User conversationUser, String connectionUrl, String webSocketTicket) {
        NextcloudTalkApplication.Companion.getSharedApplication().getComponentApplication().inject(this);

        this.connectionUrl = connectionUrl;
        this.conversationUser = conversationUser;
        this.webSocketTicket = webSocketTicket;
        this.webSocketConnectionHelper = new WebSocketConnectionHelper();
        this.usersHashMap = new HashMap<>();

        connected = false;
        eventBus.register(this);

        restartWebSocket();
    }

    private void sendHello() {
        try {
            if (TextUtils.isEmpty(resumeId)) {
                internalWebSocket.send(LoganSquare.serialize(webSocketConnectionHelper.getAssembledHelloModel(conversationUser, webSocketTicket)));
            } else {
                internalWebSocket.send(LoganSquare.serialize(webSocketConnectionHelper.getAssembledHelloModelForResume(resumeId)));
            }
        } catch (IOException e) {
            Log.e(TAG, "Failed to serialize hello model");
        }
    }

    @Override
    public void onOpen(WebSocket webSocket, Response response) {
        internalWebSocket = webSocket;
        sendHello();
    }

    private void closeWebSocket(WebSocket webSocket) {
        webSocket.close(1000, null);
        webSocket.cancel();
        if (webSocket == internalWebSocket) {
            connected = false;
            messagesQueue = new ArrayList<>();
        }

        restartWebSocket();
    }


    public void clearResumeId() {
        resumeId = "";
    }

    public void restartWebSocket() {
        reconnecting = true;

        // TODO when improving logging, keep in mind this issue: https://github.com/nextcloud/talk-android/issues/1013
        Log.d(TAG, "restartWebSocket: " + connectionUrl);
        Request request = new Request.Builder().url(connectionUrl).build();
        okHttpClient.newWebSocket(request, this);
        restartCount++;
    }

    @Override
    public void onMessage(WebSocket webSocket, String text) {
        if (webSocket == internalWebSocket) {
            Log.d(TAG, "Receiving : " + webSocket.toString() + " " + text);

            try {
                BaseWebSocketMessage baseWebSocketMessage = LoganSquare.parse(text, BaseWebSocketMessage.class);
                String messageType = baseWebSocketMessage.getType();
                switch (messageType) {
                    case "hello":
                        connected = true;
                        reconnecting = false;
                        restartCount = 0;
                        String oldResumeId = resumeId;
                        HelloResponseOverallWebSocketMessage helloResponseWebSocketMessage = LoganSquare.parse(text, HelloResponseOverallWebSocketMessage.class);
                        resumeId = helloResponseWebSocketMessage.getHelloResponseWebSocketMessage().getResumeId();
                        sessionId = helloResponseWebSocketMessage.getHelloResponseWebSocketMessage().getSessionId();
                        hasMCU = helloResponseWebSocketMessage.getHelloResponseWebSocketMessage().serverHasMCUSupport();

                        for (int i = 0; i < messagesQueue.size(); i++) {
                            webSocket.send(messagesQueue.get(i));
                        }

                        messagesQueue = new ArrayList<>();
                        HashMap<String, String> helloHasHap = new HashMap<>();
                        if (!TextUtils.isEmpty(oldResumeId)) {
                            helloHasHap.put("oldResumeId", oldResumeId);
                        } else {
                            currentRoomToken = "";
                        }

                        if (!TextUtils.isEmpty(currentRoomToken)) {
                            helloHasHap.put(ROOM_TOKEN, currentRoomToken);
                        }
                        eventBus.post(new WebSocketCommunicationEvent("hello", helloHasHap));
                        break;
                    case "error":
                        Log.e(TAG, "Received error: " + text);
                        ErrorOverallWebSocketMessage errorOverallWebSocketMessage = LoganSquare.parse(text, ErrorOverallWebSocketMessage.class);
                        if (("no_such_session").equals(errorOverallWebSocketMessage.getErrorWebSocketMessage().getCode())) {
                            Log.d(TAG, "WebSocket " + webSocket.hashCode() + " resumeID " + resumeId + " expired");
                            resumeId = "";
                            currentRoomToken = "";
                            restartWebSocket();
                        } else if (("hello_expected").equals(errorOverallWebSocketMessage.getErrorWebSocketMessage().getCode())) {
                            restartWebSocket();
                        }

                        break;
                    case "room":
                        JoinedRoomOverallWebSocketMessage joinedRoomOverallWebSocketMessage = LoganSquare.parse(text, JoinedRoomOverallWebSocketMessage.class);
                        currentRoomToken = joinedRoomOverallWebSocketMessage.getRoomWebSocketMessage().getRoomId();
                        if (joinedRoomOverallWebSocketMessage.getRoomWebSocketMessage().getRoomPropertiesWebSocketMessage() != null && !TextUtils.isEmpty(currentRoomToken)) {
                            sendRoomJoinedEvent();
                        }
                        break;
                    case "event":
                        EventOverallWebSocketMessage eventOverallWebSocketMessage = LoganSquare.parse(text, EventOverallWebSocketMessage.class);
                        if (eventOverallWebSocketMessage.getEventMap() != null) {
                            String target = (String) eventOverallWebSocketMessage.getEventMap().get("target");
                            switch (target) {
                                case "room":
                                    if (eventOverallWebSocketMessage.getEventMap().get("type").equals("message")) {
                                        Map<String, Object> messageHashMap =
                                            (Map<String, Object>) eventOverallWebSocketMessage.getEventMap().get("message");
                                        if (messageHashMap.containsKey("data")) {
                                            Map<String, Object> dataHashMap = (Map<String, Object>) messageHashMap.get(
                                                "data");
                                            if (dataHashMap.containsKey("chat")) {
                                                boolean shouldRefreshChat;
                                                Map<String, Object> chatMap = (Map<String, Object>) dataHashMap.get("chat");
                                                if (chatMap.containsKey("refresh")) {
                                                    shouldRefreshChat = (boolean) chatMap.get("refresh");
                                                    if (shouldRefreshChat) {
                                                        HashMap<String, String> refreshChatHashMap = new HashMap<>();
                                                        refreshChatHashMap.put(BundleKeys.KEY_ROOM_TOKEN, (String) messageHashMap.get("roomid"));
                                                        refreshChatHashMap.put(BundleKeys.KEY_INTERNAL_USER_ID, Long.toString(conversationUser.getId()));
                                                        eventBus.post(new WebSocketCommunicationEvent("refreshChat", refreshChatHashMap));
                                                    }
                                                }
                                            }
                                        }
                                    } else if (eventOverallWebSocketMessage.getEventMap().get("type").equals("join")) {
                                        List<HashMap<String, Object>> joinEventList = (List<HashMap<String, Object>>) eventOverallWebSocketMessage.getEventMap().get("join");
                                        HashMap<String, Object> internalHashMap;
                                        Participant participant;
                                        for (int i = 0; i < joinEventList.size(); i++) {
                                            internalHashMap = joinEventList.get(i);
                                            HashMap<String, Object> userMap = (HashMap<String, Object>) internalHashMap.get("user");
                                            participant = new Participant();
                                            String userId = (String) internalHashMap.get("userid");
                                            if (userId != null) {
                                                participant.setActorType(USERS);
                                                participant.setActorId(userId);
                                            } else {
                                                participant.setActorType(GUESTS);
                                                // FIXME seems to be not given by the HPB: participant.setActorId();
                                            }
                                            if (userMap != null) {
                                                // There is no "user" attribute for guest participants.
                                                participant.setDisplayName((String) userMap.get("displayname"));
                                            }
                                            usersHashMap.put((String) internalHashMap.get("sessionid"), participant);
                                        }
                                    }
                                    break;
                                case TARGET_PARTICIPANTS:
                                    signalingMessageReceiver.process(eventOverallWebSocketMessage.getEventMap());
                                    break;
                            }
                        }
                        break;
                    case "message":
                        CallOverallWebSocketMessage callOverallWebSocketMessage = LoganSquare.parse(text, CallOverallWebSocketMessage.class);
                        NCSignalingMessage ncSignalingMessage = callOverallWebSocketMessage.getCallWebSocketMessage().getNcSignalingMessage();
                        if (TextUtils.isEmpty(ncSignalingMessage.getFrom()) && callOverallWebSocketMessage.getCallWebSocketMessage().getSenderWebSocketMessage() != null) {
                            ncSignalingMessage.setFrom(callOverallWebSocketMessage.getCallWebSocketMessage().getSenderWebSocketMessage().getSessionId());
                        }

                        signalingMessageReceiver.process(ncSignalingMessage);
                        break;
                    case "bye":
                        connected = false;
                        resumeId = "";
                    default:
                        break;
                }
            } catch (IOException e) {
                Log.e(TAG, "Failed to recognize WebSocket message", e);
            }
        }
    }

    private void sendRoomJoinedEvent() {
        HashMap<String, String> joinRoomHashMap = new HashMap<>();
        joinRoomHashMap.put(ROOM_TOKEN, currentRoomToken);
        eventBus.post(new WebSocketCommunicationEvent("roomJoined", joinRoomHashMap));
    }

    @Override
    public void onMessage(WebSocket webSocket, ByteString bytes) {
        Log.d(TAG, "Receiving bytes : " + bytes.hex());
    }

    @Override
    public void onClosing(WebSocket webSocket, int code, String reason) {
        Log.d(TAG, "Closing : " + code + " / " + reason);
    }

    @Override
    public void onFailure(WebSocket webSocket, Throwable t, Response response) {
        Log.d(TAG, "Error : WebSocket " + webSocket.hashCode() + " onFailure: " + t.getMessage());
        closeWebSocket(webSocket);
    }

    public String getSessionId() {
        return sessionId;
    }

    public boolean hasMCU() {
        return hasMCU;
    }

    public void joinRoomWithRoomTokenAndSession(String roomToken, String normalBackendSession) {
        Log.d(TAG, "joinRoomWithRoomTokenAndSession");
        Log.d(TAG, "   roomToken: " + roomToken);
        Log.d(TAG, "   session: " + normalBackendSession);
        try {
            String message = LoganSquare.serialize(webSocketConnectionHelper.getAssembledJoinOrLeaveRoomModel(roomToken, normalBackendSession));
            if (!connected || reconnecting) {
                messagesQueue.add(message);
            } else {
                if (roomToken.equals(currentRoomToken)) {
                    sendRoomJoinedEvent();
                } else {
                    internalWebSocket.send(message);
                }
            }
        } catch (IOException e) {
            Log.e(TAG, e.getMessage(), e);
        }
    }

    private void sendCallMessage(NCSignalingMessage ncSignalingMessage) {
        try {
            String message = LoganSquare.serialize(webSocketConnectionHelper.getAssembledCallMessageModel(ncSignalingMessage));
            if (!connected || reconnecting) {
                messagesQueue.add(message);
            } else {
                internalWebSocket.send(message);
            }
        } catch (IOException e) {
            Log.e(TAG, "Failed to serialize signaling message", e);
        }
    }

    void sendBye() {
        if (connected) {
            try {
                ByeWebSocketMessage byeWebSocketMessage = new ByeWebSocketMessage();
                byeWebSocketMessage.setType("bye");
                byeWebSocketMessage.setBye(new HashMap<>());
                internalWebSocket.send(LoganSquare.serialize(byeWebSocketMessage));
            } catch (IOException e) {
                Log.e(TAG, "Failed to serialize bye message");
            }
        }
    }

    public boolean isConnected() {
        return connected;
    }

    public String getDisplayNameForSession(String session) {
        Participant participant = usersHashMap.get(session);
        if (participant != null) {
            if (participant.getDisplayName() != null) {
                return participant.getDisplayName();
            }
        }

        return "";
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    public void onMessageEvent(NetworkEvent networkEvent) {
        if (networkEvent.getNetworkConnectionEvent() == NetworkEvent.NetworkConnectionEvent.NETWORK_CONNECTED && !isConnected()) {
            restartWebSocket();
        }
    }

    public SignalingMessageReceiver getSignalingMessageReceiver() {
        return signalingMessageReceiver;
    }

    public SignalingMessageSender getSignalingMessageSender() {
        return signalingMessageSender;
    }

    /**
     * Temporary implementation of SignalingMessageReceiver until signaling related code is extracted to a Signaling
     * class.
     *
     * All listeners are called in the WebSocket reader thread. This thread should be the same as long as the
     * WebSocket stays connected, but it may change whenever it is connected again.
     */
    private static class ExternalSignalingMessageReceiver extends SignalingMessageReceiver {
        public void process(Map<String, Object> eventMap) {
            processEvent(eventMap);
        }

        public void process(NCSignalingMessage message) {
            processSignalingMessage(message);
        }
    }

    private class ExternalSignalingMessageSender implements SignalingMessageSender {
        @Override
        public void send(NCSignalingMessage ncSignalingMessage) {
            sendCallMessage(ncSignalingMessage);
        }
    }
}

File: app/src/main/java/com/nextcloud/talk/webrtc/MagicWebSocketInstance.kt
/*
 * Nextcloud Talk application
 *
 * @author Mario Danic
 * Copyright (C) 2017-2018 Mario Danic <mario@lovelyhq.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nextcloud.talk.webrtc

import android.content.Context
import android.text.TextUtils
import android.util.Log
import com.bluelinelabs.logansquare.LoganSquare
import com.nextcloud.talk.R.string
import com.nextcloud.talk.application.NextcloudTalkApplication.Companion.sharedApplication
import com.nextcloud.talk.events.NetworkEvent
import com.nextcloud.talk.events.NetworkEvent.NetworkConnectionEvent.NETWORK_CONNECTED
import com.nextcloud.talk.events.WebSocketCommunicationEvent
import com.nextcloud.talk.models.json.participants.Participant
import com.nextcloud.talk.models.json.signaling.NCMessageWrapper
import com.nextcloud.talk.models.json.websocket.*
import com.nextcloud.talk.newarch.local.models.UserNgEntity
import com.nextcloud.talk.utils.LoggingUtils.writeLogEntryToFile
import com.nextcloud.talk.utils.MagicMap
import com.nextcloud.talk.utils.bundle.BundleKeys.KEY_INTERNAL_USER_ID
import com.nextcloud.talk.utils.bundle.BundleKeys.KEY_ROOM_TOKEN
import okhttp3.OkHttpClient
import okhttp3.Request.Builder
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode.BACKGROUND
import org.koin.core.KoinComponent
import org.koin.core.inject
import java.io.IOException
import java.util.*

class MagicWebSocketInstance internal constructor(
        conversationUser: UserNgEntity,
        connectionUrl: String,
        webSocketTicket: String
) : WebSocketListener(), KoinComponent {
    val okHttpClient: OkHttpClient by inject()
    val eventBus: EventBus by inject()
    val context: Context by inject()

    private val conversationUser: UserNgEntity
    private val webSocketTicket: String
    private var resumeId: String? = null
    var sessionId: String? = null
        private set
    private var hasMCU = false
    var isConnected: Boolean
        private set
    private val webSocketConnectionHelper: WebSocketConnectionHelper
    private var internalWebSocket: WebSocket? = null
    private val magicMap: MagicMap
    private val connectionUrl: String
    private var currentRoomToken: String? = null
    private var restartCount = 0
    private var reconnecting = false
    private val usersHashMap: HashMap<String?, Participant>
    private var messagesQueue: MutableList<String> =
            ArrayList()

    private fun sendHello() {
        try {
            if (TextUtils.isEmpty(resumeId)) {
                internalWebSocket!!.send(
                        LoganSquare.serialize(
                                webSocketConnectionHelper.getAssembledHelloModel(conversationUser, webSocketTicket)
                        )
                )
            } else {
                internalWebSocket!!.send(
                        LoganSquare.serialize(
                                webSocketConnectionHelper.getAssembledHelloModelForResume(resumeId)
                        )
                )
            }
        } catch (e: IOException) {
            Log.e(TAG, "Failed to serialize hello model")
        }
    }

    override fun onOpen(
            webSocket: WebSocket,
            response: Response
    ) {
        internalWebSocket = webSocket
        sendHello()
    }

    private fun closeWebSocket(webSocket: WebSocket) {
        webSocket.close(1000, null)
        webSocket.cancel()
        if (webSocket === internalWebSocket) {
            isConnected = false
            messagesQueue = ArrayList()
        }
        restartWebSocket()
    }

    fun clearResumeId() {
        resumeId = ""
    }

    fun restartWebSocket() {
        reconnecting = true
        val request = Builder()
                .url(connectionUrl)
                .build()
        okHttpClient.newWebSocket(request, this)
        restartCount++
    }

    override fun onMessage(
            webSocket: WebSocket,
            text: String
    ) {
        if (webSocket === internalWebSocket) {
            Log.d(
                    TAG, "Receiving : $webSocket $text"
            )
            writeLogEntryToFile(
                    context,
                    "WebSocket " + webSocket.hashCode() + " receiving: " + text
            )
            try {
                val baseWebSocketMessage =
                        LoganSquare.parse(text, BaseWebSocketMessage::class.java)
                val messageType = baseWebSocketMessage.type
                when (messageType) {
                    "hello" -> {
                        isConnected = true
                        reconnecting = false
                        restartCount = 0
                        val oldResumeId = resumeId
                        val helloResponseWebSocketMessage =
                                LoganSquare.parse(
                                        text, HelloResponseOverallWebSocketMessage::class.java
                                )
                        resumeId = helloResponseWebSocketMessage.helloResponseWebSocketMessage
                                .resumeId
                        sessionId = helloResponseWebSocketMessage.helloResponseWebSocketMessage
                                .sessionId
                        hasMCU = helloResponseWebSocketMessage.helloResponseWebSocketMessage
                                .serverHasMCUSupport()
                        var i = 0
                        while (i < messagesQueue.size) {
                            webSocket.send(messagesQueue[i])
                            i++
                        }
                        messagesQueue = ArrayList()
                        val helloHasHap =
                                HashMap<String, String?>()
                        if (!TextUtils.isEmpty(oldResumeId)) {
                            helloHasHap["oldResumeId"] = oldResumeId
                        } else {
                            currentRoomToken = ""
                        }
                        if (!TextUtils.isEmpty(currentRoomToken)) {
                            helloHasHap["roomToken"] = currentRoomToken
                        }
                        eventBus.post(WebSocketCommunicationEvent("hello", helloHasHap))
                    }
                    "error" -> {
                        val errorOverallWebSocketMessage =
                                LoganSquare.parse(
                                        text, ErrorOverallWebSocketMessage::class.java
                                )
                        if ("no_such_session" ==
                                errorOverallWebSocketMessage.errorWebSocketMessage.code
                        ) {
                            writeLogEntryToFile(
                                    context,
                                    "WebSocket " + webSocket.hashCode() + " resumeID " + resumeId + " expired"
                            )
                            resumeId = ""
                            currentRoomToken = ""
                            restartWebSocket()
                        } else if ("hello_expected" ==
                                errorOverallWebSocketMessage.errorWebSocketMessage.code
                        ) {
                            restartWebSocket()
                        }
                    }
                    "room" -> {
                        val joinedRoomOverallWebSocketMessage =
                                LoganSquare.parse(
                                        text, JoinedRoomOverallWebSocketMessage::class.java
                                )
                        currentRoomToken = joinedRoomOverallWebSocketMessage.roomWebSocketMessage
                                .roomId
                        if (joinedRoomOverallWebSocketMessage.roomWebSocketMessage
                                        .roomPropertiesWebSocketMessage != null && !TextUtils.isEmpty(
                                        currentRoomToken
                                )
                        ) {
                            sendRoomJoinedEvent()
                        }
                    }
                    "event" -> {
                        val eventOverallWebSocketMessage =
                                LoganSquare.parse(
                                        text, EventOverallWebSocketMessage::class.java
                                )
                        if (eventOverallWebSocketMessage.eventMap != null) {
                            val target =
                                    eventOverallWebSocketMessage.eventMap["target"] as String?
                            when (target) {
                                "room" -> if (eventOverallWebSocketMessage.eventMap["type"] == "message"
                                ) {
                                    val messageHashMap =
                                            eventOverallWebSocketMessage.eventMap["message"] as Map<String, Any>?
                                    if (messageHashMap!!.containsKey("data")) {
                                        val dataHashMap =
                                                messageHashMap["data"] as Map<String, Any>?
                                        if (dataHashMap!!.containsKey("chat")) {
                                            val shouldRefreshChat: Boolean
                                            val chatMap =
                                                    dataHashMap["chat"] as Map<String, Any>?
                                            if (chatMap!!.containsKey("refresh")) {
                                                shouldRefreshChat = chatMap["refresh"] as Boolean
                                                if (shouldRefreshChat) {
                                                    val refreshChatHashMap =
                                                            HashMap<String, String?>()
                                                    refreshChatHashMap[KEY_ROOM_TOKEN] = messageHashMap["roomid"] as String?
                                                    refreshChatHashMap[KEY_INTERNAL_USER_ID] =
                                                            java.lang.Long.toString(conversationUser.id!!)
                                                    eventBus.post(
                                                            WebSocketCommunicationEvent("refreshChat", refreshChatHashMap)
                                                    )
                                                }
                                            }
                                        }
                                    }
                                } else if (eventOverallWebSocketMessage.eventMap["type"]
                                        == "join"
                                ) {
                                    val joinEventMap =
                                            eventOverallWebSocketMessage.eventMap["join"] as List<HashMap<String, Any>>?
                                    var internalHashMap: HashMap<String, Any>
                                    var participant: Participant
                                    var i = 0
                                    while (i < joinEventMap!!.size) {
                                        internalHashMap = joinEventMap[i]
                                        val userMap =
                                                internalHashMap["user"] as HashMap<String, Any>?
                                        participant = Participant()
                                        participant.userId = internalHashMap["userid"] as String
                                        participant.displayName = userMap!!["displayname"] as String
                                        usersHashMap[internalHashMap["sessionid"] as String?] = participant
                                        i++
                                    }
                                }
                                "participants" -> if (eventOverallWebSocketMessage.eventMap["type"] == "update"
                                ) {
                                    val refreshChatHashMap =
                                            HashMap<String, String?>()
                                    val updateEventMap =
                                            eventOverallWebSocketMessage.eventMap["update"] as HashMap<String, Any>?
                                    refreshChatHashMap["roomToken"] = updateEventMap!!["roomid"] as String?
                                    refreshChatHashMap["jobId"] = Integer.toString(
                                            magicMap.add(
                                                    updateEventMap["users"]!!
                                            )
                                    )
                                    eventBus.post(
                                            WebSocketCommunicationEvent("participantsUpdate", refreshChatHashMap)
                                    )
                                }
                            }
                        }
                    }
                    "message" -> {
                        val callOverallWebSocketMessage =
                                LoganSquare.parse(
                                        text, CallOverallWebSocketMessage::class.java
                                )
                        val ncSignalingMessage =
                                callOverallWebSocketMessage.callWebSocketMessage
                                        .ncSignalingMessage
                        if (TextUtils.isEmpty(ncSignalingMessage.from)
                                && callOverallWebSocketMessage.callWebSocketMessage.senderWebSocketMessage
                                != null
                        ) {
                            ncSignalingMessage.from =
                                    callOverallWebSocketMessage.callWebSocketMessage
                                            .senderWebSocketMessage
                                            .sessionId

                        }
                        if (!TextUtils.isEmpty(ncSignalingMessage.from)) {
                            val messageHashMap =
                                    HashMap<String, String>()
                            messageHashMap["jobId"] = Integer.toString(magicMap.add(ncSignalingMessage))
                            eventBus.post(WebSocketCommunicationEvent("signalingMessage", messageHashMap))
                        }
                    }
                    "bye" -> {
                        isConnected = false
                        resumeId = ""
                    }
                    else -> {
                    }
                }
            } catch (e: IOException) {
                writeLogEntryToFile(
                        context,
                        "WebSocket " + webSocket.hashCode() + " IOException: " + e.message
                )
                Log.e(
                        TAG, "Failed to recognize WebSocket message"
                )
            }
        }
    }

    private fun sendRoomJoinedEvent() {
        val joinRoomHashMap =
                HashMap<String, String?>()
        joinRoomHashMap["roomToken"] = currentRoomToken
        eventBus.post(WebSocketCommunicationEvent("roomJoined", joinRoomHashMap))
    }

    override fun onMessage(
            webSocket: WebSocket,
            bytes: ByteString
    ) {
        Log.d(TAG, "Receiving bytes : " + bytes.hex())
    }

    override fun onClosing(
            webSocket: WebSocket,
            code: Int,
            reason: String
    ) {
        Log.d(TAG, "Closing : $code / $reason")
        writeLogEntryToFile(
                context,
                "WebSocket " + webSocket.hashCode() + " Closing: " + reason
        )
    }

    override fun onFailure(
            webSocket: WebSocket,
            t: Throwable,
            response: Response?
    ) {
        Log.d(TAG, "Error : " + t.message)
        writeLogEntryToFile(
                context,
                "WebSocket " + webSocket.hashCode() + " onFailure: " + t.message
        )
        closeWebSocket(webSocket)
    }

    fun hasMCU(): Boolean {
        return hasMCU
    }

    fun joinRoomWithRoomTokenAndSession(
            roomToken: String,
            normalBackendSession: String?
    ) {
        try {
            val message = LoganSquare.serialize(
                    webSocketConnectionHelper.getAssembledJoinOrLeaveRoomModel(
                            roomToken,
                            normalBackendSession
                    )
            )
            if (!isConnected || reconnecting) {
                messagesQueue.add(message)
            } else {
                if (roomToken == currentRoomToken) {
                    sendRoomJoinedEvent()
                } else {
                    internalWebSocket!!.send(message)
                }
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun sendCallMessage(ncMessageWrapper: NCMessageWrapper) {
        try {
            val message = LoganSquare.serialize(
                    webSocketConnectionHelper.getAssembledCallMessageModel(ncMessageWrapper)
            )
            if (!isConnected || reconnecting) {
                messagesQueue.add(message)
            } else {
                internalWebSocket!!.send(message)
            }
        } catch (e: IOException) {
            writeLogEntryToFile(
                    context,
                    "WebSocket sendCalLMessage: " + e.message + "\n" + ncMessageWrapper.toString()
            )
            Log.e(
                    TAG, "Failed to serialize signaling message"
            )
        }
    }

    fun getJobWithId(id: Int): Any? {
        val copyJob = magicMap[id]
        magicMap.remove(id)
        return copyJob
    }

    fun requestOfferForSessionIdWithType(
            sessionIdParam: String,
            roomType: String
    ) {
        try {
            val message = LoganSquare.serialize(
                    webSocketConnectionHelper.getAssembledRequestOfferModel(sessionIdParam, roomType)
            )
            if (!isConnected || reconnecting) {
                messagesQueue.add(message)
            } else {
                internalWebSocket!!.send(message)
            }
        } catch (e: IOException) {
            writeLogEntryToFile(
                    context,
                    "WebSocket requestOfferForSessionIdWithType: "
                            + e.message
                            + "\n"
                            + sessionIdParam
                            + " "
                            + roomType
            )
            Log.e(TAG, "Failed to offer request")
        }
    }

    fun sendBye() {
        if (isConnected) {
            try {
                val byeWebSocketMessage = ByeWebSocketMessage()
                byeWebSocketMessage.type = "bye"
                byeWebSocketMessage.bye = HashMap()
                internalWebSocket!!.send(LoganSquare.serialize(byeWebSocketMessage))
            } catch (e: IOException) {
                Log.e(TAG, "Failed to serialize bye message")
            }
        }
    }

    fun getDisplayNameForSession(session: String?): String {
        return if (usersHashMap.containsKey(session)) {
            usersHashMap[session]!!.displayName
        } else sharedApplication!!.getString(string.nc_nick_guest)
    }

    fun getSessionForUserId(userId: String): String? {
        for (session in usersHashMap.keys) {
            if (userId == usersHashMap[session]!!.userId) {
                return session
            }
        }
        return ""
    }

    fun getUserIdForSession(session: String?): String {
        return if (usersHashMap.containsKey(session)) {
            usersHashMap[session]!!.userId
        } else ""
    }

    @Subscribe(threadMode = BACKGROUND)
    fun onMessageEvent(networkEvent: NetworkEvent) {
        if ((networkEvent.networkConnectionEvent
                        == NETWORK_CONNECTED) && !isConnected
        ) {
            restartWebSocket()
        }
    }

    companion object {
        private const val TAG = "MagicWebSocketInstance"
    }

    init {
        this.connectionUrl = connectionUrl
        this.conversationUser = conversationUser
        this.webSocketTicket = webSocketTicket
        webSocketConnectionHelper = WebSocketConnectionHelper()
        usersHashMap =
                HashMap()
        magicMap = MagicMap()
        isConnected = false
        eventBus.register(this)
        restartWebSocket()
    }
}